lib/index.js

```diff
@@ -0,0 +1,237 @@
+// none
+
+// variables and functions
+const defaults = {
+  // use this object to map vCard field names to editableField names. A mapping to "false" will ignore the vCard field while parsing to JSON
+  vCardToJSONAttributeMapping: {
+    BEGIN: false,
+    END: false,
+    VERSION: false,
+  },
+  // use this object to define complex vCard fields; meaning those whos technical values consist more than one logical value
+  complexJSONAttributes: {},
+};
+
+const extAttrSubvaluesMapping = {
+  'VALUE=X_EXTENSION_KEY': 'key',
+  'VALUE=X_EXTENSION_VALUE': 'value',
+  'VALUE=X_EXTENSION_DATA_TYPE': 'dataType',
+};
+
+const keyValuePairRegex = /^([^:]+):(.*)/;
+
+function VCardParser(options = {}) {
+  const toJSON = _.merge(
+    {},
+    defaults.vCardToJSONAttributeMapping,
+    options.vCardToJSONAttributeMapping
+  );
+  const complexJSONAttributes = _.merge(
+    {},
+    defaults.complexJSONAttributes,
+    options.complexJSONAttributes
+  );
+  const toVcard = _.transform(toJSON, (result, JSONAttrName, vCardAttrName) => {
+    if (_.isString(JSONAttrName)) {
+      Object.assign(result, {
+        [JSONAttrName]: vCardAttrName,
+      });
+    }
+
+    return result;
+  }, {});
+
+  Object.assign(this, {
+    complexJSONAttributes,
+    mappings: {
+      toJSON,
+      toVcard,
+    },
+  });
+}
+
+/*
+ * @ description: Parses a vCard 2.1 string and creates a JSON-Object with editableField names as properties containing the values from according vCardAttributes
+ * @ param: {string} vCardStr - The vCard string
+ *
+ * @ return {json}: a structured JSON-Object containing the corresponding profile attributes
+ *
+ * @ author: Benjamin Kroeger < benjamin.kroeger@de.ibm.com >
+ */
+VCardParser.prototype.toObject = function toObject(vCardStr, encode) {
+  const self = this;
+  const { mappings: { toJSON }, complexJSONAttributes } = self;
+  const jsonObj = {
+    extattr: [],
+  };
+  const trimmedVCardStr = vCardStr.trim();
+  // split the string on "\n" to receive singular vCardAttributes
+  const validVCardAttributes = [];
+
+  // see if the vCardAttribute starts with one of the configured JSONAttribute-mappings
+  trimmedVCardStr.split('\n').forEach((vCardLine) => {
+    let valid = false;
+    _.forOwn(toJSON, (JSONAttrName, vCardAttrName) => { // eslint-disable-line consistent-return
+      if (vCardLine.indexOf(vCardAttrName) === 0) {
+        valid = true;
+        if (_.isString(JSONAttrName) && JSONAttrName.length > 0) {
+          validVCardAttributes.push(vCardLine);
+        }
+        // break from the loop
+        return false;
+      }
+    });
+    if (
+      !valid &&
+      _.isString(validVCardAttributes[validVCardAttributes.length - 1])
+    ) {
+      // no mapping was found; either it has not been configured or the value from the previous vCardAttribute contained a "\n". To prevent from data-loss, value is appended to the previous attribute.
+      validVCardAttributes[validVCardAttributes.length - 1] += ` ${vCardLine}`;
```

Comment:
This line is also a potential bug. What if that last element in this array is a string, and the `vCardLine` is not valid. This line of code appends invalid value to a valid one, which can cause problems in the next parsing steps.

What do you think about adding a debug message, that says:
`'Please provide valid mapping for [${vCardLine}]'`

And then, a developer can add a new attribute to `options.vCardToJSONAttributeMapping` that supports this line as well.

Reply:
how can the vcard line not be valid? vcard line is by definition a string (might be empty) since it resulted from `vCardStr.split('\n')`.
This logic was added since it might happen that a vcard attribute value includes line breaks (`\n`)... and we don't want to throw away parts of the value.
One example is the `aboutMe` and `background` attributes from IBM Connections Profiles. Both of which include line breaks